Refactorings notifications

sascha11110 opened the pull request:
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I documented the TypeScript code using JSDoc style.

### Motivation and Context
During the development of some new features for the notification system, multiple refactorings have become necessary. In order to keep the the feature branches compact, I separated the refactorings to this branch.

### Description
- Added `SingleUserNotificationFactory` and `GroupNotificationFactory` to clean up `SingleUserNotificationService` and `GroupNotificationService`.
- Cleaned up all domain classes related to notifications
- All components related to `notification-management` are solely used for managing system notifications. So I renamed all files accordingly.
- Used `notification.class` in `NotificationRepository` to distinguish the notification types more consistently.
- Cleaned up the placement of translation strings related to notifications
- Added attributes/columns `priority` and `oudated` to the notification model ([Liquibase changelog](https://github.com/ls1intum/Artemis/blob/02c9daa595e5d782d458a7dda5d2c5450a92715f/src/main/resources/config/liquibase/changelog/20200426110000_changelog.xml))

The changes to the data model go along with further planned changes which I have documented in my proposal with the following Analysis Object Model:

![AOM](https://user-images.githubusercontent.com/15168372/81474376-52ced580-9205-11ea-8899-bc5dfcb6838c.png)

### Steps for Testing
#### Trigger notification
1. Log in to Artemis (at least as tutor)
2. Go to any exercise where the Q&A feature is activated
3. Create a new question
4. Make sure you get a notification for the created question

#### System Notifications
1. Log in to Artemis
2. Click *Server Administration* in the top navigation and click on *System notifications* in the submenu
3. Create, update, delete system notifications and check that everything works as expected

#### Translation strings
Go to the following views and check that translation strings are correct:
- Notification sidebar: Click notification bell in top navigation
- Notification overview: Click notification bell in top navigation -> click on *Show all notifications*
- System notifications: Click *Server administration* in top navigation -> click on *System notifications* in the sub-menu (edit, detail, delete)

Check with english and with german language setting.

### Screenshots
No UI changes.

madwau commented on src/main/java/de/tum/in/www1/artemis/domain/notification/GroupNotificationFactory.java:
```diff
@@ -0,0 +1,162 @@
+            text = "Attachment \"" + attachment.getName() + "\" updated.";
+        }
+        else {
+            throw new UnsupportedOperationException("Unsupported NotificationType.");
```
Maybe include `notificationType` in the error message so it's clear which notification type is unsupported and caused the error.

sascha11110 replied:
Good point, I will add it!

madwau commented on src/main/java/de/tum/in/www1/artemis/domain/notification/GroupNotificationFactory.java:
```diff
@@ -0,0 +1,162 @@
+        Course course = attachment.getLecture().getCourse();
+        GroupNotification notification = new GroupNotification(course, title, text, author, groupNotificationType);
+
+        // Set target
```
This comment doesn't really add much value since it's just the method name. I'd say leave it out or expand more.

sascha11110 replied:
I will remove it :-)

madwau commented on src/main/java/de/tum/in/www1/artemis/domain/notification/GroupNotificationFactory.java:
```diff
@@ -0,0 +1,162 @@
+     */
+    public static GroupNotification createNotification(Exercise exercise, User author, GroupNotificationType groupNotificationType, NotificationType notificationType) {
+        String title, text;
+        if (notificationType == NotificationType.EXERCISE_CREATED) {
```
This seems like a good opportunity for a switch statement.

sascha11110 replied:
Normally I prefer if-else over switch statements in Java but the new enhanced switch statements or really nice (multiple values possible, no break in every case needed). I will adjust my implementation :-)

madwau commented on src/main/java/de/tum/in/www1/artemis/service/GroupNotificationService.java:
```diff
@@ -7,220 +7,145 @@
+     */
+    public void notifyStudentGroupAboutAttachmentChange(Attachment attachment, String notificationText) {
+        // Do not send a notification before the release date of the attachment.
+        if (attachment.getReleaseDate() != null && !attachment.getReleaseDate().isBefore(ZonedDateTime.now())) {
```
I think this would be easier to parse mentally:

```suggestion
        if (attachment.getReleaseDate() != null && attachment.getReleaseDate().isAfter(ZonedDateTime.now())) {
```

sascha11110 replied:
Definitely!

madwau commented on src/main/java/de/tum/in/www1/artemis/service/GroupNotificationService.java:
```diff
@@ -7,220 +7,145 @@
      *
      * @param exercise that has been updated
-     * @param notificationText that should be displayed
+     * @param notificationText that should be displayed  TODO: can be removed in the future as notification's text attribute is not used in the client
```
Could this be removed now or are there difficulties associated with removing it?

sascha11110 replied:
Forgot to work on that. Actually, we need to keep the notification text as there is an option to set a custom notification message when updating an attachment or an exercise.

madwau commented on src/main/java/de/tum/in/www1/artemis/service/GroupNotificationService.java:
```diff
@@ -7,220 +7,145 @@
+     * @param notificationText that should be displayed  TODO: can be removed in the future as notification's text attribute is not used in the client
+     */
+    public void notifyInstructorGroupAboutExerciseUpdate(Exercise exercise, String notificationText) {
+        // TODO: why null for author?
```
Who is this question directed at? Maybe it'd be better to clarify the question since the `TODO` adds more confusion than it alleviates, I feel.

sascha11110 replied:
I forgot this and the previous TODO. Was somehow a reminder for myself. I will resolve both todos.

madwau commented on src/main/webapp/app/admin/system-notification-management/system-notification-management-detail.component.html:
```diff
@@ -0,0 +1,23 @@
+        <dt><span jhiTranslate="artemisApp.systemNotification.type">Type</span></dt>
+        <dd>{{ notification.type }}</dd>
+        <dt><span jhiTranslate="artemisApp.systemNotification.expireDate">Expire Date</span></dt>
+        <dd>{{ notification.expireDate | date: 'yyyy-MM-dd HH:mm' }}</dd>
```
Is this date format wanted here? Maybe it makes sense in this scenario but I think in general we wanted to go with `dd-MM-yyyy HH:mm` everywhere on the platform.

sascha11110 replied:
I just took it from the existing implementation of notification-component.html. I will adjust it in both places.

sascha11110 replied:
I will use `dd.MM.yyyy HH:mm` as it seems to be the most common format in the code base.

madwau commented on src/main/webapp/app/admin/system-notification-management/system-notification-management-detail.component.html:
```diff
@@ -0,0 +1,23 @@
+<div *ngIf="notification">
+    <h2>
+        <span jhiTranslate="artemisApp.systemNotification.systemNotification">System notification</span> [<b>{{ notification.id }}</b
+        >]
```
The line break looks a bit odd.

sascha11110 replied:
Good catch! Prettier introduced that line break. I split it up in two lines.

madwau commented on src/main/webapp/app/admin/system-notification-management/system-notification-management-detail.component.ts:
```diff
@@ -6,9 +6,9 @@ import { SystemNotificationService } from 'app/core/system-notification/system-n
+    templateUrl: './system-notification-management-detail.component.html',
 })
-export class NotificationMgmtDetailComponent implements OnInit, OnDestroy {
+export class SystemNotificationMgmtDetailComponent implements OnInit, OnDestroy {
```
Should we switch `Mgmt` to `Management`? I think the overall codebase tends to favor verbosity over abbreviations.

sascha11110 replied:
I will change that :-)

madwau commented on src/main/webapp/i18n/en/notification.json:
```diff
@@ -0,0 +1,38 @@
+            "title": "Title",
+            "text": "Text",
+            "type": "Type",
+            "expireDate": "Expire Date",
```
```suggestion
            "expireDate": "Expiration Date",
```

madwau commented:
Unfortunately I couldn't get the first test case to work. I didn't receive a notification when a question was created (I once created it using the same account and once I created the question with a student account). The logged in user is in all of the course groups.

The second test case worked as expected. One small detail I noticed: It would be nice to only show the colon `:` if the notification has any text:

![OnPaste 20200505-145422](https://user-images.githubusercontent.com/7109225/81068232-50583d00-8ee0-11ea-917c-2ec809b4ec46.png)

The third test case was also fine; translations all looked good to me :+1:

sascha11110 commented on src/main/java/de/tum/in/www1/artemis/service/GroupNotificationService.java:
```diff
@@ -7,220 +7,145 @@
+            return;
+        }
+        // Create and send the notification.
+        saveAndSend(GroupNotificationFactory.createNotification(attachment, userService.getUser(), GroupNotificationType.STUDENT, NotificationType.ATTACHMENT_CHANGE));
```
Didn't know that this possibility exists in Java. I will add it :-)

sascha11110 commented on src/main/java/de/tum/in/www1/artemis/domain/notification/GroupNotificationFactory.java:
```diff
@@ -0,0 +1,162 @@
+import de.tum.in.www1.artemis.domain.enumeration.GroupNotificationType;
+import de.tum.in.www1.artemis.domain.enumeration.NotificationType;
+
+public class GroupNotificationFactory {
```
I'm actually planning that as a part of the upcoming outdated mechanism for the different notification types (e.g. newAnswerForExercise, exerciseUpdate, ...). But I think for now this is a little bit out of scope for this PR :-)

jpbernius replied:
Sure, just an idea.  :D

sascha11110 commented on src/main/webapp/app/admin/system-notification-management/system-notification-management.route.ts:
```diff
@@ -0,0 +1,70 @@
+    }
+}
+
+export const systemNotificationMgmtRoutes1: Route = {
```
I have simply stuck to the previous implementation. I will adjust that.

sascha11110 commented:
@jpbernius Thanks for testing! I answered all and incorporated most of your suggested changes.

> * From your "Steps for Testing - System Notifications":
>   > Click Server Administration in the top navigation and click on Notification Management in the submenu
>   > In the client, it says "System notifications", is this expected?

I've created the PR description before I changed it to "System notifications". I adjusted the PR description.

> * If I delete a notification, it will only disappear after reloading the page.
> * A scheduled notification will not just appear once the time passes (I think this is expected?)

Both problems were unfortunately already there before. I will fix that in a follow-up PR (see also https://github.com/ls1intum/Artemis/issues/1328).

> * "Steps for Testing - Translation strings:
>   All notification texts are English, regardless of my language setting.

Can't reproduce that. Does this also apply to other views like the student's course overview? Then it might be related to your account or local setup.

sascha11110 commented:
@madwau Thanks for testing! Especially for your comprehensive code review!

I answered all and incorporated most of your suggested changes. Unfortunately I can't use the new switch statements as stated in one comment because spotless does not fully support them yet.

> Unfortunately I couldn't get the first test case to work. I didn't receive a notification when a question was created (I once created it using the same account and once I created the question with a student account). The logged in user is in all of the course groups.

I wasn't able to reproduce that. On the test server and locally. Maybe we can have together a look at it?

> The second test case worked as expected. One small detail I noticed: It would be nice to only show the colon : if the notification has any text:

I fixed that :-)
